### PR TITLE
Create Zero Tolerance Plugin

### PR DESCRIPTION
**Pull request by SerperiorBae**
This plugin was taken from Events.js and repurposed for the Marketplace Staff & Marketplace Rooms. It was added due to the fact that we currently have to use a rudimentary zero tolerance pastebin; I do plan to look into opening this command into all rooms, so that each room can have its list of zero tolerance users that they can keep track of easily and efficiently.
- If events.js works, this will work, have tested.

### CODE REVIEW

File: chat-plugins/zerotolerance.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+					return this.sendReply("There is already an User named: " + params[0] + " on the Zero Tolerance List.");
+				}
+			}
+			ZeroTolUsers["Name"] = toId(params[0]);
+			ZeroTolUsers["DisplayName"] = params[0];
+			ZeroTolUsers["Reason"] = params[1];
+			saveUsers();
+		},
+		remove: function (target, room, user) {
+			if (!this.can('mute', null, room)) return this.errorReply('/0tol - Access denied.');
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:57]: this isn't even the command

File: chat-plugins/zerotolerance.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+			}
+			let display = '<center><strong style="font-size: 20px; font-weight: bold;">Zero Tolerance Users</strong><br><table style="' + TABLE_STYLE + '"><tr><th style="' + OUTER_TH_STYLE + '">User Name</th><th style="' + INNER_TH_STYLE + '">Reason</th>';
+			for (let zerotol in UserList) {
+				display += '<tr><td style="' + TD_STYLE + '">' + UserList[zerotol]["DisplayName"] + '</td><td style="' + TD_STYLE + '">' + UserList[zerotol]["Reason"] + '</td></tr><tr>';
+			}
+			this.sendReply('|raw|' + display);
+		},
+	},
+	zerotolhelp: ['Event Command by Skyla (Full)',
+	'- /0tol create [name]| [description] - Creates and stores an entry in the Zero Tolerance Database.',
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:94]: this isn't the command

File: chat-plugins/zerotolerance.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	'- /0tol remove [name] - Removes user from Zero Tolerance Database.',
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:95]: same here

File: chat-plugins/zerotolerance.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	'- /0tol display - Displays all Zero Tolerance Users.'],
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:96]: and here

File: chat-plugins/zerotolerance.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+				for (let zerotol in UserList) {
+					if (UserList[zerotol]["Name"] === params || UserList[zerotol]["DisplayName"] === params) {
+						this.sendReply("The User: \"" + params + "\" has been removed from the Zero Tolerance List.");
+						delete UserList[zerotol];
+						fs.writeFileSync('config/zerotol.json', JSON.stringify(UserList));
+						break;
+					}
+				}
+			} else {
+				return this.sendReply("/0tol - The User: \"" + params + "\" is not on the Zero Tolerance List");
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:72]: and here

File: chat-plugins/zerotolerance.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	UserList[count] = ZeroTolUsers;
+	fs.writeFileSync('config/zerotol.json', JSON.stringify(UserList));
+}
+exports.commands = {
+	zerotol: {
+		create: function (target, room, user) {
+			let params = target.split("| ") || target.split("|");
+			if (!this.can('mute', null, room)) return this.errorReply('/0tol - Access denied.');
+			if (room.id !== 'marketplace' || room.id !== 'marketplacestaff') return this.errorReply('This command can only be used in Marketplace Staff');
+			if (!target || !params[0] || !params[1]) return this.errorReply('Usage: /0tol add [Name]| [Reason]');
```
Comment (panpawn) [inline chat-plugins/zerotolerance.js:38]: and here

**Comment by panpawn** [comment]
Also, I'm pretty sure that main is working on blacklisted users, server side, so I don't know if I see a purpose for this system with that in mind... https://github.com/Zarel/Pokemon-Showdown/pull/2545

**Comment by SerperiorBae** [comment]
Oh no panpawn, this is like completely different, like zerotolerance in marketplace isn't "Blacklisted" it's just where we up the punishments for the users for their behaviour and is one step behind being banned / blacklisted.

**Comment by panpawn** [comment]
So what's the point then?

**Comment by panpawn** [comment]
I don't see enough of a purpose of this command.

**Pull request closed without merging**